llvm/lib/SYCLLowerIR/SYCLDeviceRequirements.cpp

```diff
@@ -0,0 +1,136 @@
+  MDString *S = cast<llvm::MDString>(N->getOperand(OpNo).get());
+  return S->getString();
+}
+
+SYCLDeviceRequirements
+llvm::getSYCLDeviceRequirements(const module_split::ModuleDesc &MD) {
+  SYCLDeviceRequirements Reqs;
+  // Process all functions in the module
+  for (const Function &F : MD.getModule()) {
+    if (auto *MDN = F.getMetadata("sycl_used_aspects")) {
+      for (size_t I = 0, E = MDN->getNumOperands(); I < E; ++I) {
+        auto Val = ExtractSignedIntegerFromMDNodeOperand(MDN, I);
+        // Don't put internal aspects (with negative integer value) into the
+        // requirements, they are used only for device image splitting.
+        if (Val >= 0)
+          Reqs.Aspects.insert(Val);
+      }
+    }
+
+    if (auto *MDN = F.getMetadata("sycl_fixed_targets")) {
+      for (size_t I = 0, E = MDN->getNumOperands(); I < E; ++I) {
+        auto Val = ExtractUnsignedIntegerFromMDNodeOperand(MDN, I);
+        Reqs.FixedTarget.insert(Val);
+      }
+    }
+
+    if (auto *MDN = F.getMetadata("reqd_work_group_size")) {
+      llvm::SmallVector<uint64_t, 3> NewReqdWorkGroupSize;
+      for (size_t I = 0, E = MDN->getNumOperands(); I < E; ++I)
+        NewReqdWorkGroupSize.push_back(
+            ExtractUnsignedIntegerFromMDNodeOperand(MDN, I));
+      if (!Reqs.ReqdWorkGroupSize.has_value())
+        Reqs.ReqdWorkGroupSize = NewReqdWorkGroupSize;
+    }
+
+    if (auto *MDN = F.getMetadata("sycl_joint_matrix")) {
+      auto Val = ExtractStringFromMDNodeOperand(MDN, 0);
+      if (!Val.empty())
+        Reqs.JointMatrix = Val;
+    }
+
+    if (auto *MDN = F.getMetadata("sycl_joint_matrix_mad")) {
+      auto Val = ExtractStringFromMDNodeOperand(MDN, 0);
+      if (!Val.empty())
+        Reqs.JointMatrixMad = Val;
+    }
+  }
+
+  // Process just the entry points in the module
+  for (const Function *F : MD.entries()) {
+    if (auto *MDN = F->getMetadata("intel_reqd_sub_group_size")) {
+      // There should only be at most one function with
+      // intel_reqd_sub_group_size metadata when considering the entry
+      // points of a module, but not necessarily when considering all the
+      // functions of a module: an entry point with a
+      // intel_reqd_sub_group_size can call an ESIMD function through
+      // invoke_esimd, and that function has intel_reqd_sub_group_size=1,
+      // which is valid.
+      assert(
+          MDN->getNumOperands() == 1 &&
+          "intel_reqd_sub_group_size metadata expects exactly one argument!");
+      auto MDValue = ExtractUnsignedIntegerFromMDNodeOperand(MDN, 0);
+      if (!Reqs.SubGroupSize)
+        Reqs.SubGroupSize = MDValue;
+      else
+        assert(*Reqs.SubGroupSize == static_cast<uint32_t>(MDValue));
+    }
+  }
+  return Reqs;
+}
+
+std::map<StringRef, util::PropertyValue> SYCLDeviceRequirements::asMap() const {
+  std::map<StringRef, util::PropertyValue> Requirements;
+
+  Requirements["aspects"] =
+      std::vector<uint32_t>(Aspects.begin(), Aspects.end());
```

Comment:
What if `Aspects` set is empty? Shouldn't we apply `if (<property is set>) <fill Requirements>` here as it's done for other properties?

Reply:
I was preserving the behavior of the code before I refactored this part, which always attached the property. I believe the runtime can handle the property not being attach but I also might need to update some tests if I do.

https://github.com/intel/llvm/blob/ea7ba1b965302277fc23ef48dba83b10e6c734e9/llvm/tools/sycl-post-link/SYCLDeviceRequirements.cpp#L64-L68